### DISCUSSION TITLE
Add a rotated image to another

### DISCUSSION THREAD

**Discussion by poms32** [ImageMagick/ImageMagick#6622] [Help] [answered]
### ImageMagick version

7.1.1-11

### Operating system, version and so on

Mac Os Ventura 13.5.1

### Description

Hi there !

I searched a lot on the legacy forum and here but did not find any viable solution for my problem.

The goal is to add an image in another image. BUT, I need to rotate the added image so I came up with something like that : 

`convert prod.png \
\( img.png -resize 1984.5x2604 -rotate 320  \) \
-gravity NorthWest -geometry -716.37177-1374.2147 -composite out.png`

My problem is that my geometry coordinates are for the top left corner of the rotated image considered as à "layer" but in ImageMagick after being rotated the new rotated image is bigger than the original one and the top left corner is not the same as the original one. (If I am being clear)

So my question is how to guess/get the top left coordinates of the rotated image in order to update the geometry to the right place. Or how can I handle the rotation and placement in another way ?

Thanks a lot for your help

**Comment by fmw42** [comment]
You might compute your geometry center to center and then use -gravity center

**Comment by snibgo** [comment]
> (If I am being clear)

You have described what happens, and said this isn't what you want.

But you haven't said what you do want. Perhaps you want the pixel that was at top-left of img.png to be in the top-left position in the output, as if img.png was initially placed at the top-left position and then rotated around its top-left corner.  Then, of course, some of img.png will not be in the output.

Or perhaps you want something else. I suggest you post sample inputs, and a mock-up (perhaps made in Gimp) of the output you want.

**Comment by poms32** [comment]
Hello both of you and thank you for your reply.

Here is a mockup of what I am working with :
![CleanShot 2023-09-07 at 08 36 55](https://github.com/ImageMagick/ImageMagick/assets/8656129/9901f4b2-bb16-4d68-b82e-173995d3aff6)

the goal of my command is to combine these two images knowing this data. My problem is that the given coordinates of the top left corner of img.png is after the rotation of r° but ImageMagick is not working in the same way, so when rotate img.png and set its coordinates the positionning is wrong.

I'll try @fmw42 idea to work with the center and let you know if I managed to do it.

Thank you

**Comment by snibgo** [comment]
How about:
```
magick -size 500x500 xc:Blue ( toes.png -virtual-pixel None +distort SRT 1,320 ) -compose Over -layers Flatten x.png
```
![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/ade6a65d-d890-43aa-8c84-6b233ac81437)

**Comment by poms32** [comment]
The rotation is not the problem actually. My problem is to place the added rotated image at the right place.

In my example mockup I want to place the rotated image at 30 from the top and -10 from the left but this is for the top left corner of the added image. And this is the only informationI have is this.

My guess would be to calculate the top left corner of the rotated image (as shown below in green) but I don't know how to do it, or maybe there is another way ?

![CleanShot 2023-09-07 at 14 31 13](https://github.com/ImageMagick/ImageMagick/assets/8656129/f69bee41-ca2c-46be-b86b-4a21f39f5ff5)

Thank you

**Reply by fmw42** [reply]
Can you post an example pair of images and define how you want it placed relative to the top left corner.  One of us can give you a formula for placement.

**Comment by GeeMack** [comment, marked as answer]
ImageMagick's  "-distort" operator can rotate an image from a specific point, then place that point at a specified location in the output. So to turn the overlay image 320 degrees *and* place the original point 0,0 at the -10,30 position... 

    magick prod.png  img.png -background none -extent %[fx:u.w]x%[fx:u.h] ^
       +repage ( +clone -distort SRT "0,0 1 320 -10,30" ) -delete 1 -composite result.png

That reads both images into the command, then extends the canvas of the overlay image to the dimensions of the background image. Next it isolates a clone of the overlay inside parentheses, and uses a "-distort SRT" operation to rotate and move the overlay to it's final position on the canvas. After the parentheses it deletes the un-rotated overlay image, and composites the modified overlay onto the background.

The command is in Windows CMD syntax. For a Windows BAT script double the percent signs "%%". For a *nix OS change the continued-line caret "^" to a backslash "\\", escape the parentheses with backslashes "\\(...\\)", and maybe put double quotes around FX expressions "%[fx:...]" if there are special characters for your shell.

**Reply by poms32** [reply]
This actually works pretty good ! I am making some more test and will mark as answered !

Thank you very much for your help

**Reply by poms32** [reply]
The xeird thing is that it gives me a full black background.. Am I missing something ?

FYI the prod.png is a full transparent png file.

**Comment by snibgo** [comment]
> ... it gives me a full black background.

Are you saving it to a format that can't record transparency, like JPG?

With "-distort", we often want "-virtual-pixel None" before it.

If that doesn't help, then please show your input files, and the exact command you used.

**Comment by poms32** [comment]
I am really sorry it was totally my fault, I changed to out.jpg and forget to revert my code. So it works fine on my server too. 

Again sorry for the loss of time. the answer given by @GeeMack is working fine.

Thank you all for your time and help.

Have a good day
